v2/api/dataprotection/customizations/backup_vaults_backup_instance_extensions.go

```diff
@@ -0,0 +1,142 @@
+ * Licensed under the MIT license.
+ */
+
+package customizations
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-service-operator/v2/internal/set"
+	"github.com/Azure/azure-service-operator/v2/internal/util/to"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	armdataprotection "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/dataprotection/armdataprotection/v3"
+	dataprotection "github.com/Azure/azure-service-operator/v2/api/dataprotection/v1api20231101/storage"
+
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/extensions"
+)
+
+var (
+	_ extensions.PreReconciliationChecker  = &BackupVaultsBackupInstanceExtension{}
+	_ extensions.PostReconciliationChecker = &BackupVaultsBackupInstanceExtension{}
+)
+
+var nonRetryableStates = set.Make(
+	"ConfiguringProtectionFailed",
+	"Invalid",
+	"NotProtected",
+	"ProtectionConfigured",
+	"SoftDeleted",
+	"ProtectionStopped",
+	"BackupSchedulesSuspended",
+	"RetentionSchedulesSuspended",
+)
+
+var retryableStates = set.Make(
+	"ProtectionError",
+	"UpdatingProtection",
+	"ConfiguringProtection",
+	"SoftDeleting",
+)
+
+func (extension *BackupVaultsBackupInstanceExtension) PostReconcileCheck(
+	ctx context.Context,
+	obj genruntime.MetaObject,
+	owner genruntime.MetaObject,
+	_ *resolver.Resolver,
+	armClient *genericarmclient.GenericClient,
+	log logr.Logger,
+	_ extensions.PostReconcileCheckFunc,
+) (extensions.PostReconcileCheckResult, error) {
+	log.V(Debug).Info("########################## Starting Post-reconcilation for Backup Instance ##########################")
+	backupInstance, ok := obj.(*dataprotection.BackupVaultsBackupInstance)
+	if !ok {
+		return extensions.PostReconcileCheckResult{},
+			errors.Errorf("cannot run on unknown resource type %T, expected *dataprotection.BackupVaultsBackupInstance", obj)
+	}
+
+	var _ conversion.Hub = backupInstance
+
+	if owner != nil && backupInstance != nil && backupInstance.Status.Id != nil &&
+		backupInstance.Status.Properties != nil && backupInstance.Status.Properties.ProtectionStatus != nil &&
+		backupInstance.Status.Properties.ProtectionStatus.Status != nil {
+
+		protectionStatus := *backupInstance.Status.Properties.ProtectionStatus.Status
+		log.V(Debug).Info(fmt.Sprintf("########################## Protection Status is  %q ##########################", protectionStatus))
+
+		// return success for reconcilation if the state is non-retryable
+		if nonRetryableStates.Contains(protectionStatus) {
+			log.V(Debug).Info("########################## Returning PostReconcileCheckResultSuccess ##########################")
+			return extensions.PostReconcileCheckResultSuccess(), nil
```

Comment:
This is where you should be calling the passed `extensions.PostReconcileCheckFunc`, but only if the resource is in a non-error condition - which I think means only if the state is `ProtectionConfigured`, but I'm not 100% on that.

Reply:
I am not able to see any example for this or places where its been done. Please point me to the same.

Reply:
The `next` parameter is provided to give the extension flexibility to augment or replace functionality as it chooses - it's a more powerful approach than providing a `PostReconcileCheck` that then always does the standard thing regardless.
 
The most usual approach is to handle special cases, then delegate to `next()` for the standard cases:

``` go
if specialCaseApplies() {
    // custom behaviour
} else {
    next()
}
```

Sometimes, an extension wants to do something after the standard behaviour has executed:

``` go
result := next()
if specialCaseApplies(result) {
    // Override the result
}

return result
```

Very very rarely, an extension wants to completely take over and skip the default behaviour. 

``` go
// Don't call 
// next()
SpecialCase()
```

In this last case, the extension author is stating "I know exactly what to do here, this will always be correct, _even if the default behaviour changes in the future_".

There are cases where this is true. We have some in ASO. But they aren't a precedent for other extensions skipping next().

Unless you have a really compelling reason otherwise, you should always call `next()`. 

I have a TODO task to write up more information about the extensions, and this guidance will be a significant part of it.